Deprecated WebSecurityConfigurerAdapter in spring-security 5.7.1

BoykoAlex opened the pull request:
See: https://spring.io/blog/2022/02/21/spring-security-without-the-websecurityconfigureradapter

TODO is `configure(AuthenticationManagerBuilder)` - it is left for a later contribution

BoykoAlex commented:
@pway99 I have pushed the revised version of the Recipe to play it safe when non applicable methods are overridden and/or used within the same class declaration.

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+            }
+
+            private boolean isApplicableClass(J.ClassDeclaration classDecl) {
+                if (classDecl.getExtends() != null) {
+                    JavaType.FullyQualified superType = TypeUtils.asFullyQualified(classDecl.getExtends().getType());
+                    if (superType != null && FQN_WEB_SECURITY_CONFIGURER_ADAPTER.equals(superType.getFullyQualifiedName())
+                            && classDecl.getLeadingAnnotations().stream().anyMatch(a -> TypeUtils.isOfClassType(a.getType(), FQN_CONFIGURATION))) {
```
Consider simplifying this with `TypeUtils#isAssignableTo`
```suggestion
                if (TypeUtils.isAssignableTo(FQN_WEB_SECURITY_CONFIGURER_ADAPTER, classDecl.getType())
                    && classDecl.getLeadingAnnotations().stream().anyMatch(a -> TypeUtils.isOfClassType(a.getType(), FQN_CONFIGURATION))) {
```

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+
+                            @Override
+                            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, AtomicBoolean found) {
+                                if (getCursor().firstEnclosing(J.ClassDeclaration.class) == classDecl) {
```
here you can use `Cursor#isScopeInPath`
```java
if (getCursor().isScopeInPath(classDecl)) {
```

pway99 replied:
here you can use `Cursor#isScopeInPath`
```java
if (getCursor().isScopeInPath(classDecl)) {
```

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+
+                            @Override
+                            public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, AtomicBoolean found) {
+                                if (getCursor().firstEnclosing(J.ClassDeclaration.class) == classDecl) {
```
here you can use Cursor#isScopeInPath

pway99 replied:
here you can use Cursor#isScopeInPath

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+
+                            @Override
+                            public J.MemberReference visitMemberReference(J.MemberReference memberRef, AtomicBoolean found) {
+                                if (getCursor().firstEnclosing(J.ClassDeclaration.class) == classDecl) {
```
here you can use `Cursor#isScopeInPath`

pway99 replied:
here you can use `Cursor#isScopeInPath`

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+
+                            private void checkNonApplicableMethod(JavaType.Method m, AtomicBoolean found) {
+                                if (!found.get()) {
+                                    if (WEB_SECURITY_CONFIGURER_ADAPTER_METHOD_MATCHER.matches(m)
```
The `WEB_SECURITY_CONFIGURER_ADAPTER_METHOD_MATCHER` matches any `FQN_WEB_SECURITY_CONFIGURER_ADAPTER` method, consider checking the methods declaring type instead.

```java
if (TypeUtils.isAssignableTo(FQN_WEB_SECURITY_CONFIGURER_ADAPTER, m.getDeclaringType())
```

pway99 replied:
The `WEB_SECURITY_CONFIGURER_ADAPTER_METHOD_MATCHER` matches any `FQN_WEB_SECURITY_CONFIGURER_ADAPTER` method, consider checking the methods declaring type instead.

```java
if (TypeUtils.isAssignableTo(FQN_WEB_SECURITY_CONFIGURER_ADAPTER, m.getDeclaringType())
```

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+    private static final String FQN_BEAN = BEAN_PKG + "." + BEAN_SIMPLE_NAME;
+    private static final String BEAN_ANNOTATION = "@" + BEAN_SIMPLE_NAME;
+
+    private static final MethodMatcher WEB_SECURITY_CONFIGURER_ADAPTER_METHOD_MATCHER =
+            new MethodMatcher(FQN_WEB_SECURITY_CONFIGURER_ADAPTER + " *(..)", true);
```
this is unnecessary if `TypeUtils#isAssignableTo` is used for checking the methods declaring type.

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+                    J.ClassDeclaration classDecl = classDeclCursor.getValue();
+                    if (classDeclCursor.getMessage(MSG_QUALIFIES, false)) {
+                        if (CONFIGURE_HTTP_SECURITY_METHOD_MATCHER.matches(parentMethod, classDecl)) {
+                            J.VariableDeclarations.NamedVariable paramDef = ((J.VariableDeclarations) parentMethod.getParameters().get(0)).getVariables().get(0);
+                            J.Identifier methodName = new J.Identifier(Tree.randomId(), Space.EMPTY, Markers.EMPTY, "build", null, null);
+                            JavaType.Method methodType = TypeUtils.findDeclaredMethod(TypeUtils.asFullyQualified(paramDef.getType()), "build", Collections.emptyList())
+                                    .orElse(new JavaType.Method(null, 0, null, "build", null, null, null, null, null));
+                            J.Identifier variableIdentifier = new J.Identifier(Tree.randomId(), Space.EMPTY, Markers.EMPTY, paramDef.getSimpleName(), paramDef.getType(), null);
+                            J.MethodInvocation invocation = new J.MethodInvocation(Tree.randomId(), Space.build(" ", Collections.emptyList()), Markers.EMPTY, JRightPadded.build(variableIdentifier), null, methodName, JContainer.build(Collections.emptyList()), methodType);
+                            J.Return returnStatement = new J.Return(Tree.randomId(), Space.build("\n", Collections.emptyList()), Markers.EMPTY, invocation);
+                            b = b.withStatements(ListUtils.concat(b.getStatements(), autoFormat(returnStatement, context)));
+//                        JavaTemplate template = JavaTemplate.builder(() -> getCursor(), "return " + ((J.VariableDeclarations)parentMethod.getParameters().get(0)).getVariables().get(0).getSimpleName() + ".build();").javaParser(() -> JavaParser.fromJavaVersion().build()).build();
+//                        b = b.withTemplate(template, b.getCoordinates().addStatement(null));
```
your JavaTemplate was super close it just needed a parameter reference and a symbol for the SecurityBuilder.
check this out:

```java
JavaTemplate template = JavaTemplate.builder(this::getCursor,  "return #{any(org.springframework.security.config.annotation.SecurityBuilder)}.build();")
                                   .javaParser(() -> JavaParser.fromJavaVersion()
                                           .dependsOn("package org.springframework.security.config.annotation;" +
                                                   "public interface SecurityBuilder<O> {\n" +
                                                   "    O build() throws Exception;" +
                                                   "}")
                                           .build()).imports("org.springframework.security.config.annotation.SecurityBuilder").build();
                            b = b.withTemplate(template, b.getCoordinates().lastStatement(), ((J.VariableDeclarations)parentMethod.getParameters().get(0)).getVariables().get(0).getName());
```

BoykoAlex replied:
Thanks very much! I completely forgot about template parameters. The example solution above is a great reminder of this feature!-) Will do my best not to forget this next time!-) Thanks again!

pway99 replied:
sure any time, The JavaTemplate can be a bit tricky to configure ;)

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,240 @@
+                    J.ClassDeclaration classDecl = classDeclCursor.getValue();
+                    if (classDeclCursor.getMessage(MSG_QUALIFIES, false)) {
+                        if (CONFIGURE_HTTP_SECURITY_METHOD_MATCHER.matches(parentMethod, classDecl)) {
+                            J.VariableDeclarations.NamedVariable paramDef = ((J.VariableDeclarations) parentMethod.getParameters().get(0)).getVariables().get(0);
+                            J.Identifier methodName = new J.Identifier(Tree.randomId(), Space.EMPTY, Markers.EMPTY, "build", null, null);
+                            JavaType.Method methodType = TypeUtils.findDeclaredMethod(TypeUtils.asFullyQualified(paramDef.getType()), "build", Collections.emptyList())
+                                    .orElse(new JavaType.Method(null, 0, null, "build", null, null, null, null, null));
+                            J.Identifier variableIdentifier = new J.Identifier(Tree.randomId(), Space.EMPTY, Markers.EMPTY, paramDef.getSimpleName(), paramDef.getType(), null);
+                            J.MethodInvocation invocation = new J.MethodInvocation(Tree.randomId(), Space.build(" ", Collections.emptyList()), Markers.EMPTY, JRightPadded.build(variableIdentifier), null, methodName, JContainer.build(Collections.emptyList()), methodType);
+                            J.Return returnStatement = new J.Return(Tree.randomId(), Space.build("\n", Collections.emptyList()), Markers.EMPTY, invocation);
+                            b = b.withStatements(ListUtils.concat(b.getStatements(), autoFormat(returnStatement, context)));
+//                        JavaTemplate template = JavaTemplate.builder(() -> getCursor(), "return " + ((J.VariableDeclarations)parentMethod.getParameters().get(0)).getVariables().get(0).getSimpleName() + ".build();").javaParser(() -> JavaParser.fromJavaVersion().build()).build();
+//                        b = b.withTemplate(template, b.getCoordinates().addStatement(null));
```
your JavaTemplate was super close it just needed a parameter reference and a symbol for the SecurityBuilder.
check this out:

```java
JavaTemplate template = JavaTemplate.builder(this::getCursor,  "return #{any(org.springframework.security.config.annotation.SecurityBuilder)}.build();")
                                   .javaParser(() -> JavaParser.fromJavaVersion()
                                           .dependsOn("package org.springframework.security.config.annotation;" +
                                                   "public interface SecurityBuilder<O> {\n" +
                                                   "    O build() throws Exception;" +
                                                   "}")
                                           .build()).imports("org.springframework.security.config.annotation.SecurityBuilder").build();
                            b = b.withTemplate(template, b.getCoordinates().lastStatement(), ((J.VariableDeclarations)parentMethod.getParameters().get(0)).getVariables().get(0).getName());
```

BoykoAlex commented:
@pway99 Thanks for the review and comments! Thanks very much again for the template with parameter solution!

I have addressed all you remarks/corrections except one:
```
getCursor().firstEnclosing(J.ClassDeclaration.class) == classDecl
```
replace with
```
getCursor().isScopeInPath(classDecl)
```
I'd want the first enclosing class to find non-applicable methods in my class and not inner classes. I have added a test for this case. Thus kept `firstEnclosing(...)`

BoykoAlex commented:
Hmm... on the second though if it's no n static inner class then method invocations from the outer WebSecurityConfigurerAdapter class make it non-applicable...

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,230 @@
+                            b = b.withTemplate(template, b.getCoordinates().lastStatement(),
+                                    ((J.VariableDeclarations)parentMethod.getParameters().get(0)).getVariables().get(0).getName());
+                        } else if (CONFIGURE_WEB_SECURITY_METHOD_MATCHER.matches(parentMethod, classDecl)) {
+                            J.Lambda.Parameters parameters = new J.Lambda.Parameters(Tree.randomId(), Space.EMPTY, Markers.EMPTY, true, parentMethod.getParameters().stream().filter(J.VariableDeclarations.class::isInstance).map(J.VariableDeclarations.class::cast).map(p -> {
+                                J.VariableDeclarations.NamedVariable varDecl = p.getVariables().get(0);
+                                return new JRightPadded<J>(varDecl.getName(), Space.EMPTY, Markers.EMPTY);
+                            }).collect(Collectors.toList()));
+
+                            J.Lambda lambda = new J.Lambda(Tree.randomId(), Space.build(" ", Collections.emptyList()), Markers.EMPTY, parameters, Space.build(" ", Collections.emptyList()), b.getStatements().size() == 1 ? b.getStatements().get(0).withPrefix(Space.EMPTY) : b, null);
+                            J.Return returnStatement = new J.Return(Tree.randomId(), Space.build("\n", Collections.emptyList()), Markers.EMPTY, lambda);
+                            b = b.withStatements(Collections.singletonList(autoFormat(returnStatement, context)));
```
this could be converted to use a java template,  the result is slightly different but functionally equivalent


```java
                            String t = "return (" + ((J.VariableDeclarations)parentMethod.getParameters().get(0)).getVariables().get(0).getName().getSimpleName() + ") -> #{any()};";
                            JavaTemplate template = JavaTemplate.builder(this::getCursor, t).javaParser(() -> JavaParser.fromJavaVersion().build()).doBeforeParseTemplate(System.out::println).build();
                            b = b.withTemplate(template, b.getCoordinates().firstStatement(), b);
                            b = b.withStatements(ListUtils.map(b.getStatements(), (index, stmt) -> {
                                    if (index == 0){
                                        return stmt;
                                    }
                                    return null;
                                }));
```
result:
```java
                    return (web) -> {
                        web.ignoring().antMatchers("/ignore1", "/ignore2");
                    };
```

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,230 @@
+                maybeAddImport(FQN_BEAN);
+                JavaTemplate template = JavaTemplate.builder(() -> c, BEAN_ANNOTATION).imports(FQN_BEAN).javaParser(() -> JavaParser.fromJavaVersion()
+                        .dependsOn("package " + BEAN_PKG + "; public @interface " + BEAN_SIMPLE_NAME + " {}").build()).build();
+                return m.withTemplate(template, m.getCoordinates().addAnnotation(null));
```
```suggestion
                return m.withTemplate(template, m.getCoordinates().addAnnotation(Comparator.comparing(J.Annotation::getSimpleName)));
```

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,230 @@
+                                    }
+                                    return anno;
+                                }))
+                                .withReturnTypeExpression(new J.Identifier(Tree.randomId(), m.getReturnTypeExpression().getPrefix(), Markers.EMPTY,"SecurityFilterChain", securityChainType, null))
```
`m.getReturnTypeExpression().getPrefix()` may throw an NPE

pway99 commented on src/main/java/org/openrewrite/java/spring/boot2/WebSecurityConfigurerAdapterRecipe.java:
```diff
@@ -0,0 +1,230 @@
+                                }))
+                                .withMethodType(type)
+                                .withParameters(Collections.emptyList())
+                                .withReturnTypeExpression(new J.Identifier(Tree.randomId(), m.getReturnTypeExpression().getPrefix(), Markers.EMPTY,"WebSecurityCustomizer", securityCustomizerType, null))
```
`m.getReturnTypeExpression().getPrefix()` may throw an NPE

pway99 commented:
This is really good work! 
A few JavaTemplate changes and some minor suggestions.
Thanks so much for your help @BoykoAlex

BoykoAlex commented:
@pway99 Thanks again for the review! I have updated the PR with your latest suggestions and corrections.

pway99 commented:
Thanks @BoykoAlex :)
